File: slices/gatekeeper/gatekeeper-core/src/internal/error-pages.ts
type OAuthErrorKind =
  | 'unsupported_response_type'
  | 'unsupported_code_challenge'
  | 'invalid_redirect_uri'
  | 'invalid_scheme'
  | 'unknown_client'
  | 'disabled_client'
  | 'redirect_uri_not_allowed'
  | 'scope_not_allowed'

const escape = (value: string): string =>
  value.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')

const oauthErrorMessage: Record<OAuthErrorKind, { title: string; body: string }> = {
  unsupported_response_type: {
    title: 'Unsupported response type',
    body: 'Only the authorization code flow (response_type=code) is supported',
  },
  unsupported_code_challenge: {
    title: 'Unsupported code challenge method',
    body: 'Only S256 code_challenge_method is supported',
  },
  invalid_redirect_uri: {
    title: 'Invalid redirect URI',
    body: 'The supplied redirect_uri is not a well-formed URL.',
  },
  invalid_scheme: {
    title: 'Invalid redirect URI scheme',
    body: 'The supplied redirect_uri must use http or https.',
  },
  unknown_client: {
    title: 'Unknown client',
    body: 'The supplied client_id is not registered.',
  },
  disabled_client: {
    title: 'Disabled client',
    body: 'The supplied client_id has been disabled.',
  },
  redirect_uri_not_allowed: {
    title: 'Redirect URI not allowed',
    body: 'The supplied redirect_uri is not registered for this client.',
  },
  scope_not_allowed: {
    title: 'Scope not allowed',
    body: 'One or more requested scopes are not permitted for this client.',
  },
}

const renderErrorHtml = (title: string, body: string): string =>
  `<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>${title}</title>
</head>
<body>
  <h1>${title}</h1>
  <p>${body}</p>
</body>
</html>`

const oauthErrorHtml = (kind: OAuthErrorKind, received?: string): string => {
  const { title, body } = oauthErrorMessage[kind]
  let suffix = ''
  if (
    (kind === 'unsupported_code_challenge' || kind === 'unsupported_response_type') &&
    received != null
  ) {
    suffix = ` (received: ${escape(received)})`
  }
  return renderErrorHtml(title, `${body}${suffix}.`)
}

export { oauthErrorHtml }

File: slices/gatekeeper/gatekeeper-rust/src/error_pages.rs
pub enum OAuthErrorKind {
    UnsupportedCodeChallenge,
    InvalidRedirectUri,
    InvalidScheme,
    UnknownClient,
    DisabledClient,
    RedirectUriNotAllowed,
    ScopeNotAllowed,
}

fn escape(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
}

fn title_and_body(kind: &OAuthErrorKind) -> (&'static str, &'static str) {
    match kind {
        OAuthErrorKind::UnsupportedCodeChallenge => (
            "Unsupported code challenge method",
            "Only S256 code_challenge_method is supported",
        ),
        OAuthErrorKind::InvalidRedirectUri => (
            "Invalid redirect URI",
            "The supplied redirect_uri is not a well-formed URL.",
        ),
        OAuthErrorKind::InvalidScheme => (
            "Invalid redirect URI scheme",
            "The supplied redirect_uri must use http or https.",
        ),
        OAuthErrorKind::UnknownClient => (
            "Unknown client",
            "The supplied client_id is not registered.",
        ),
        OAuthErrorKind::DisabledClient => (
            "Disabled client",
            "The supplied client_id has been disabled.",
        ),
        OAuthErrorKind::RedirectUriNotAllowed => (
            "Redirect URI not allowed",
            "The supplied redirect_uri is not registered for this client.",
        ),
        OAuthErrorKind::ScopeNotAllowed => (
            "Scope not allowed",
            "One or more requested scopes are not permitted for this client.",
        ),
    }
}

pub fn oauth_error_html(kind: OAuthErrorKind, method: Option<&str>) -> String {
    let (title, body) = title_and_body(&kind);
    let suffix = match (&kind, method) {
        (OAuthErrorKind::UnsupportedCodeChallenge, Some(m)) => format!(" (received: {})", escape(m)),
        _ => String::new(),
    };
    format!(
        "<!doctype html>\n\
<html lang=\"en\">\n\
<head>\n\
  <meta charset=\"utf-8\">\n\
  <title>{title}</title>\n\
</head>\n\
<body>\n\
  <h1>{title}</h1>\n\
  <p>{body}{suffix}.</p>\n\
</body>\n\
</html>"
    )
}
